### PR TITLE
pull request -- Snow Leopards Bukunmi G

### PR DESCRIPTION
**Pull request by bukunmig**
Thank You

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ def test_adds_movie_to_user_watchlist():
     assert updated_data["watchlist"][0]["genre"] == GENRE_1
     assert updated_data["watchlist"][0]["rating"] == RATING_1
 
-@pytest.mark.skip()
+
 def test_moves_movie_from_watchlist_to_empty_watched():
```
Comment (anselrognlie) [inline tests/test_wave_01.py:104]: To get this test to pass we need to remove the `raise Exception("Test needs to be completed.")` and also add additional `assert` statements to ensure that the expected behavior has taken place. For instance, here we would like to make sure that the movie dict that was in the watchlist list is now in the watched list.

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -124,7 +124,7 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
 
-@pytest.mark.skip()
+
 def test_moves_movie_from_watchlist_to_watched():
```
Comment (anselrognlie) [inline tests/test_wave_01.py:128]: To get this test to pass we need to remove the `raise Exception("Test needs to be completed.")` and also add additional `assert` statements to ensure that the expected behavior has taken place. Here, we would like to make sure that the movie dict that was in the watchlist list is now in the watched list, and that the other movies remained in their lists unchanged.

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ def test_friends_unique_movies():
     assert FANTASY_4 in friends_unique_movies
     assert amandas_data == clean_wave_3_data()
 
-@pytest.mark.skip()
+
 def test_friends_unique_movies_not_duplicated():
```
Comment (anselrognlie) [inline tests/test_wave_03.py:48]: This test also has an Exception needs to be removed, and then you'll want to add additional assertions to ensure that the function is working as expected.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,86 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    movie_dict = {'title': title, 'genre' : genre, 'rating': rating}
+    if title and genre and rating:
+        return movie_dict
+    else:
+        return None
```
Comment (anselrognlie) [inline viewing_party/party.py:8]: Consider rearranging the code flow here to first check for any of the error-like conditions, and do the more simple action of returning `None` before we even try to make our dictionary. Something like:
```py
    if not title or not genre or not rating:
        return None

    movie_dict = {'title': title, 'genre' : genre, 'rating': rating}
    return movie_dict
```

This puts the error-like check at the top, after which we don't need to think about the error cases. Then our main code can be unindented, which helps emphasize that it's the main focus of the function.

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,86 @@
...
 
 def create_movie(title, genre, rating):
-    pass
+    movie_dict = {'title': title, 'genre' : genre, 'rating': rating}
+    if title and genre and rating:
+        return movie_dict
+    else:
+        return None
+    
+def add_to_watched(user_data,movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:10]: 👍

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,86 @@
...
+    user_data['watched'].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data,movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:15]: 👍

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,86 @@
...
+
+def watch_movie(user_data,title): 
+    movie_exist = False
+    i = 0
+
+    for movie in user_data['watchlist']:
+        if movie['title']==title:
+            
+            user_data['watched'].append(user_data['watchlist'].pop(i))
+            return user_data
```
Comment (anselrognlie) [inline viewing_party/party.py:28]: Nice use of return immediately after modifiying the list. This can help us ensure that we don't try to perform multiple modifications on the list, which we saw in roundtable can have unexpected results.

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,86 @@
...
+            movie_exist = True
+    i+=1
```
Comment (anselrognlie) [inline viewing_party/party.py]: This should be indented under the `for` loop so that we increment `i` in tandem with comparing the movie title. As written, `i` wouldn't be modified until the for loop completes, which means `i` remains `0` throughout the loop, and the code always tries to pop the movie at position `0`, no matter which movie was actually found to match the title.

I completed the wave 1 tests locally, and this increment being out of place was causing the one of them (`test_moves_movie_from_watchlist_to_watched`) to still fail.

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,86 @@
...
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    rating = 0.0
+    if not user_data['watched']:
+        return 0.0
```
Comment (anselrognlie) [inline viewing_party/party.py:45]: 👍 Nice handling of when the list is empty. By checking this early and returning, it means we don't have to do any further special handling later. And because the if branch returns, the further logic doesn't need to be under an `else:`, making it more readable.

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,86 @@
...
+    for movie in user_data['watched']:
+        rating +=movie['rating']
 
+    average_rating = rating/len(user_data['watched'])
```
Comment (anselrognlie) [inline viewing_party/party.py:49]: 👍

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,86 @@
...
+    return average_rating
+
+def get_most_watched_genre(user_data):
+    counter =0 
+    genre_list = []
+    genre_dict = {}
+    if not user_data['watched']:
+        return None
```
Comment (anselrognlie) [inline viewing_party/party.py:57]: 👍 Another great "guard clause" to check for a simple case and return a simple result.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,86 @@
...
+    for movie in user_data['watched']:
+        genre_list.append(movie.get('genre'))
```
Comment (anselrognlie) [inline viewing_party/party.py:59]: We wouldn't expect a movie to not have a genre key. We only need to use `.get` if we expect there's a chance that the requested key won't be in the dictionary. But all our movies should have a genre.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,86 @@
...
+    i = 0
+    genre_loc = 0
+    for genre in genre_list:
+        genre_count = genre_list.count(genre)
```
Comment (anselrognlie) [inline viewing_party/party.py:63]: The `.count` function can be handy, but how would we tally this up ourselves? We should also be aware that the `.count` function conceals a hidden loop within, leading to this being a O(n^2) way of counting up the times we see each genre.

**Review by anselrognlie — COMMENTED** [review]
This is a good start.

Most of the tests for waves 1 and 2 are passing, and they all would be with a small tweak. I included feedback on those waves, so please review that and then let's get together to look over the rest of the project!

File: tests/test_wave_01.py [hunk 14]
```diff
@@ -143,12 +143,12 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    assert len(updated_data["watched"]) !=len(updated_data["watchlist"])
```
Comment (anselrognlie) [inline tests/test_wave_01.py:150]: Notice that from the assertions on 143,144 this must already be true, since they check that one has a length of 1, and the other 2. The question is _which_ movies are in the watchlist list and _which_ movies are in the watched list.

Given the test setup, which is the movie expect to still be in the watchlist list, and which 2 movies should now be in the watched list. Then how can we check that those movies are where we expect them to be. Do we have variables in the test referring to the movies we need to verify? How can we check whether those values are in a list in python?

File: tests/test_wave_03.py [hunk 15]
```diff
@@ -1,8 +1,9 @@
+from turtle import title
```
Comment (anselrognlie) [inline tests/test_wave_03.py]: 👀 Watch out for stray imports that VSCode pulls in when being "helpful" during autocomplete. There's a [post in study hall](https://adad18.slack.com/archives/C03QJHJDW2V/p1663446293671619?thread_ts=1663446288.654079&cid=C03QJHJDW2V) about how to turn this VSCode behavior off.

File: tests/test_wave_03.py [hunk 16]
```diff
@@ -55,12 +56,12 @@ def test_friends_unique_movies_not_duplicated():
     # Assert
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
     # **************************************************************************************************
+    # assert ["friends"]["watched"]["title"] != [amandas_data]["watched"]["title"]
```
Comment (anselrognlie) [inline tests/test_wave_03.py]: 👀 We need to think about what the expected collection of movies in the friends list might be. We can go look at the data definitions for the content used in the test and figure out what the unique colelction of movies should be, and then write the code that looks for those movies in the result.

We might also notice that this test is identical to the test above it, except that it adds a duplicate movie to one of the friends.

File: tests/test_wave_05.py [hunk 17]
```diff
@@ -38,7 +38,7 @@ def test_new_genre_rec_from_empty_watched():
     # Assert
     assert len(recommendations) == 0
 
-@pytest.mark.skip()
+# @pytest.mark.skip()
 def test_new_genre_rec_from_empty_friends():
```
Comment (anselrognlie) [inline tests/test_wave_05.py:42]: 👀 This test would need to be finsihed to complete wave 5.

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,165 @@
...
+    # for movie in user_data["watched"]:
+    #     if movie not in user_data["friends"]["watched"]:
+    #         unique_list.append(movie)
+    # return unique_list
+    all_friends_movies = []
+    i = 0
+    for friend in user_data["friends"]:
+        for friend_movie in user_data["friends"][i]["watched"]:
+            all_friends_movies.append(friend_movie)
+        i += 1
```
Comment (anselrognlie) [inline viewing_party/party.py:87]: Nice idea to build a combined list of all the movies the friends have watched! But since you're iterating over the friends list with a `for/in` loop, we can avoid using indexes to keep track of where we are in the list. `friend` will refer to each friend dict in the list as it iterates.

```py
    for friend in user_data["friends"]:
        for friend_movie in friend["watched"]:
            all_friends_movies.append(friend_movie)
```

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,165 @@
...
+
+    friend_titles = []
+    for friend_movie in all_friends_movies:
+        friend_titles.append(friend_movie["title"])
```
Comment (anselrognlie) [inline viewing_party/party.py:91]: Great idea to focus on the movie titles to use for checking whether the user has seen the movie. But since you're building this to eventually perform `in` checks on it, consider using a `set` instead, which has a much more efficient `in` check than a list does.

```py
    friend_titles = set()
    for friend_movie in all_friends_movies:
        friend_titles.add(friend_movie["title"])
```

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,165 @@
...
+            friends_not_watched.append(user_movie)
+    return friends_not_watched
+
+def get_friends_unique_watched(user_data):
+    unwatched_list = []
+    counter = 0
+    for movie in user_data ["friends"]: 
+        for friend_movie in user_data["friends"][counter]["watched"]:
+            unwatched_list.append(friend_movie)
+        counter +=1
```
Comment (anselrognlie) [inline viewing_party/party.py:105]: The friends list contains friend records, so `friend` would be a more descriptive name for the loop variable. Like in `get_unique_watched`, we don't need to count how far along in the list we are since we're iterating useing a `for/in`

```py
    for friend in user_data ["friends"]: 
        for friend_movie in friend["watched"]:
            unwatched_list.append(friend_movie)
```

File: viewing_party/party.py [hunk 21]
```diff
@@ -1,23 +1,165 @@
...
+
+    seen_movies = set()
+    clean_list = []
+    for friends_movie in unwatched_list:
+        t = tuple(friends_movie.items())
```
Comment (anselrognlie) [inline viewing_party/party.py:110]: We need to be careful using the `items()` helper to get the dict contents for packing into a tuple. We don't know that for every dictionary with the same key/value pairs, that they are in the same "order", and when comparing tuples, order matters. The underlying idea is great, but we might want to make a helper function to ensure that the values are are always read in the same order (such as by ordering the keys alphabetically).

Alternatively, as you did for `get_unique_watched` you could use the movie title as a stand-in for the entire movie.

File: viewing_party/party.py [hunk 22]
```diff
@@ -1,23 +1,165 @@
...
+        if t not in seen_movies:
+            seen_movies.add(t)
+            clean_list(friends_movie)
```
Comment (anselrognlie) [inline viewing_party/party.py]: 👀 The `.append` is missing.

```py
            clean_list.append(friends_movie)

```

File: viewing_party/party.py [hunk 23]
```diff
@@ -1,23 +1,165 @@
...
+    i = 0
+    for friend in user_data["friends"]:
+        for friends_movie in user_data["friends"][i]["watched"]:
+            all_friends_movies.append(friends_movie)
+        i += 1
+
+    friends_and_user = []
+    for friend_movies in all_friends_movies:
+        if friend_movies not in user_data["watched"]:
+            friends_and_user.append(friend_movies)
```
Comment (anselrognlie) [inline viewing_party/party.py:138]: Notice lines 128-138 are really getting the movies that the firends have watched that the user has not. So we could replace this with a call to `get_friends_unique_watched`

```py
    friends_and_user = get_friends_unique_watched(user_data)
```

File: viewing_party/party.py [hunk 24]
```diff
@@ -1,23 +1,165 @@
...
+
+    recomended_movies = []
+    j = 0
+    for movie in friends_and_user:
+        for subscription in user_data["subscriptions"]:
+            if subscription in friends_and_user[j]["host"]:
+                recomended_movies.append(movie)
+        j += 1
```
Comment (anselrognlie) [inline viewing_party/party.py:146]: 👀 Another place we can avoid using indices. We also don't need to iterate over the subscriptions (we can do an `in` check on the subscription list). But be careful of what you put on the left and right of an `in` check. The existing code works (line 144) because `in` with a string and a string does a substring check, and since equal strings are substrings of each other, in ends up not mattering.

But if we want to see whether a string is in a list of strings, then we need to but the individual string to the left, and the list to the right.

```py
    for movie in friends_and_user:
        if movie["host"] in user_data["subscriptions"]:
            recomended_movies.append(movie)
```

File: viewing_party/party.py [hunk 25]
```diff
@@ -1,23 +1,165 @@
...
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+rec_genre_list =[]
+    
+def user_genre(user_data):
+    return max(set(user_data["watched"]), key = user_data["watched"].count)
```
Comment (anselrognlie) [inline viewing_party/party.py:158]: The watched list is a list of dictionaries, which cannot go into a set. While this set of commands _can_ get the count of the number of (immutable) items in a list, it's important to understand how it works if we're going to use code like this. How would you explain it's operation to a classmate?

But also remember that we already have a function that can calculate the user's most watched genre already!

File: viewing_party/party.py [hunk 26]
```diff
@@ -1,23 +1,165 @@
...
+
+def get_new_rec_by_genre(user_data):
+    for movie in (user_data):
+        if movie in user_data["friends"]["watched"] and movie not in user_data['watched']:
```
Comment (anselrognlie) [inline viewing_party/party.py:162]: The friends list is a list of friend dictionaries, so we would need to check for the movie in each friend's watched list. But it looks like you might be trying to get the movies that the friends have watched, but the user hasn't. We've already written a function that does this.

Could we use our two previous functions to simplify the implementation of _this_ function?

**Review by anselrognlie — COMMENTED** [review]
You've made some great progress! Let's discuss some of the remaining issues from my feedback during our 1:1.

File: tests/test_wave_01.py [hunk 27]
```diff
@@ -143,12 +143,15 @@ def test_moves_movie_from_watchlist_to_watched():
...
+    assert FANTASY_2 in updated_data["watched"]
+    assert HORROR_1 in updated_data["watched"]
+    assert FANTASY_1 in updated_data["watchlist"]
```
Comment (anselrognlie) [inline tests/test_wave_01.py:153]: Great! We discussed how to think about the assertions, and how the existing assertions were checking the lengths, but that we still needed to validate the movies in the lists were the ones we expected.

File: tests/test_wave_03.py [hunk 28]
```diff
@@ -54,13 +55,16 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
+    assert INTRIGUE_3 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert FANTASY_4 in friends_unique_movies
```
Comment (anselrognlie) [inline tests/test_wave_03.py:60]: 👍 Here as well, we explored the input data, reasoned about what should be in hte result ourselves, then wrote the checks to ensure that the 3 movies in the liist are the expected ones.

File: viewing_party/party.py [hunk 29]
```diff
@@ -1,23 +1,165 @@
...
+            unwatched_list.append(friend_movie)
+        counter +=1
+
+    seen_movies = set()
+    clean_list = []
+    for friends_movie in unwatched_list:
+        t = tuple(friends_movie.items())
+        if t not in seen_movies:
+            seen_movies.add(t)
+            clean_list.append(friends_movie)
```
Comment (anselrognlie) [inline viewing_party/party.py:113]: 👍 Just that little issue. And we discussed trying to avoid using indexing so much while iterating over a list.

**Review by anselrognlie — COMMENTED** [review]
👍 Great job updating your submission. We talked through a few of the most important comments. Please be sure to review the remaining comments and be sure to ask me if there's anything unclear. Project and PSE comments are the most important feedback for improving your coding, so if I didn't explain my suggestion very well, be sure to ask me what I meant!

But great job working through this! 🎉
